### PR TITLE
Support Configuration File

### PR DESCRIPTION
**Pull request by pantor**
This PR adds the support of a configuration file, as described in the updated Readme. This way, checks can be enabled/disabled (fixing #64) and the code style can be adapted to specific needs.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,8 +37,46 @@ you can lint every file in the current directory. Using the directory `d`-flag,
...
+disable-by-inline-comments=true
+
+[line-length]
+max-line-length=120
+
+[note]
+keywords=TODO,FIXME
+```
+
+In the *Checks* group, each check can be enabled/disabled individually. The *Disabler* group allows for disabling a single check at a specific line using an inline comment (see Disabling Errors below). Furthermore, each check can have individual - hohpefully self-explanatory - settings, which are also listed in the [wiki](https://github.com/elementary/vala-lint/wiki/Vala-Lint-Checks).
```
Comment (jeremypw) [inline README.md]: individual, hopefully self-explanatory, settings

**Comment by jeremypw** [comment]
Since the default is to have all checks enabled, would it not be simpler just to have those that are disabled (or more generally not what is the default) in the config file?  Any check not in the config file would then get its default setting rather than throwing an error.  This would make it more robust if an old config file were used after new checks were added to vala-lint.

**Comment by jeremypw** [comment]
Is it possible to have a singleton Config class rather having a new instance per check?

**Comment by jeremypw** [comment]
Is it necessary to add all the checks to global checks and then filter those that are disabled out again?  Why not only add those that are enabled?  I guess you would need some kind of CheckFactory class that could take the check name and return the corresponding Check object if it is enabled or null if not.  You could then loop through a list of check names and add the enabled ones.  You could leave this for another PR though.

**Comment by pantor** [comment]
Currently, if settings are missing the config file, the default values will be used. The config shown in the Readme or generated by `--generate-config` are just the default values.

**Comment by jeremypw** [comment]
There doesn't appear to be a default value for the  `enabled` property of the Check class or its sub-classes. Won't it be set to `false` if config.get_boolean () fails because the key is missing?

**Comment by pantor** [comment]
The default values are set in the `Config.vala` file. But I'm not sure if that's a good code style, a singleton class should be a step in the right direction.

**Comment by pantor** [comment]
I've now implemented this as a singleton class, I think it is much cleaner now.

**Comment by pantor** [comment]
Is there anything I can do for this PR? I feel like support for a configuration file is important for vala-lint, but I'm unsure if my implementation is good to go.

**Comment by jeremypw** [comment]
I'll try to have a look at this in the near future.

**Comment by jeremypw** [comment]
So when vala-lint is integrate with CI. the CI will fail if there is *any* code-style error? At the moment nearly everything will fail.  Maybe the default should be "warn" for the time being? What do you think?

**Comment by kgrubb** [comment]
I've got a draft of adding in a "warn" state here: https://github.com/vala-lang/vala-lint/pull/103. Perhaps this can be augmented to set/determine the check's level?

**Comment by pantor** [comment]
@jeremypw I think we should leave the „error“ state the default but add an `—exit-zero` command line option for CI instead. This is similar to some other linters I know (e.g. pylint).

**Comment by kgrubb** [comment]
The exit-zero arg doesn't really cover all cases though. If I want notes to show up in my lint output, but also don't want to make the ci job fail due to notes, I only have 2 choices in this situation:

1. Disable notes (now they don't show up in my output though)
2. Keep notes enabled and add an --exit-zero arg (now other errors may show up and get subsequently ignored)

Ideally, I would want the notes to show up in my output, but not affect the exit code (and let other matches trigger exit codes still).

But perhaps that's something we pursue in a separate pr, built off of this one?

**Comment by pantor** [comment]
@kgrubb At the moment, there are three states (levels) ("off", "error", "warn") which you can specify in the config file for each check individually and alter the output and exit code accordingly. I think that should satisfy your issue #98 and your question above. Or am I missing something? The zero exit flag is just an additional option in particular for CI. 

The main difference in the implementation here and your PR #103 are:
 - The level is only tied to each check, and not to each format mistake.
 - The level is not printed.

I think both additions should then be done in your separate PR :wink:

**Comment by kgrubb** [comment]
No, you're not missing anything! I misunderstood the warning level changes you've added. The readme update clarified it! Just tried it out locally and it's working as expected. :tada:

**Comment by kgrubb** [comment]
Here's a screenshot of this branch in action. I set "note" to warn in the generated config for this example:

Screenshot from 2019-09-19 14-37-12

**Pull request merged**